src/async_fuse/memfs/cache/storage_manager.rs

```diff
@@ -0,0 +1,533 @@
+        data: &[u8],
+    ) -> impl Iterator<Item = IoBlock> + '_ {
+        let data_len = data.len();
+
+        let mut blocks = vec![];
+
+        // Handle the first block
+        let mut first_block = Block::new(self.block_size);
+        let first_block_offset = offset.overflow_rem(self.block_size);
+        let first_block_end = first_block_offset
+            .overflow_add(data_len)
+            .min(self.block_size);
+        let first_block_length = first_block_end.overflow_sub(first_block_offset);
+        let first_block_data = data
+            .get(..first_block_length)
+            .unwrap_or_else(|| unreachable!("`data` is checked that it has enough bytes."));
+        first_block
+            .make_mut()
+            .get_mut(first_block_offset..first_block_end)
+            .unwrap_or_else(|| {
+                unreachable!("Both `blk_offset` and `blk_end` must not be greater than `blk_size`")
+            })
+            .copy_from_slice(first_block_data);
+        blocks.push(first_block);
+
+        // Handle the rest blocks
+        if data_len > first_block_length {
+            let chunks = data
+                .get(first_block_length..)
+                .unwrap_or_else(|| unreachable!("`data` is checked that it has enough bytes."))
+                .chunks(self.block_size);
+            for chunk in chunks {
+                let chunk_len = chunk.len();
+                let mut block = Block::new(self.block_size);
+                block
+                    .make_mut()
+                    .get_mut(..chunk_len)
+                    .unwrap_or_else(|| {
+                        unreachable!("The `capacity` must be greater than `chunk_len`")
+                    })
+                    .copy_from_slice(chunk);
+                blocks.push(block);
+            }
+        }
+
+        let block_num = blocks.len();
+
+        // Convert blocks to IoBlock
+        blocks.into_iter().enumerate().map(move |(i, block)| {
+            let offset = if i == 0 { first_block_offset } else { 0 };
+
+            let end = if i == 0 {
+                first_block_end
+            } else if i == block_num.overflow_sub(1) {
+                first_block_offset
+                    .overflow_add(data_len)
+                    .overflow_sub(1)
+                    .overflow_rem(self.block_size)
+                    .overflow_add(1)
+            } else {
+                self.block_size
+            };
+
+            IoBlock::new(block, offset, end)
+        })
+    }
+
+    /// Load data from storage.
+    pub async fn load(
+        &self,
+        ino: INum,
+        offset: usize,
+        len: usize,
+        mtime: SystemTime,
+    ) -> Vec<IoBlock> {
+        // Check if the cache is valid.
+        let invalid = {
+            let guard = pin();
+            let cache_mtime = self.mtimes.get(&ino, &guard);
+            cache_mtime != Some(&mtime)
+        };
+
+        if invalid {
+            self.storage.invalidate(ino).await;
```

Comment:
It should be an async task. We need to create an issue to document the design of asynchronous tasks for the storage module. We can do this later.

Reply:
Or a flag called `valid` to reuse memory space? We could think it later. Create an issue about reusing memory.

Reply:
> It should be an async task. We need to create an issue to document the design of asynchronous tasks for the storage module. We can do this later.

But we must wait for the invalidation before we fetch a block from the storage, or we may success to load invalid blocks.

Reply:
We could add a flag called `valid` to Block (like the dirty bit). Invalid operation just set the bit to false. Then we could reuse the Block's memory.  We could consider this when you start to refactor the old system.

Reply:
> We could add a flag called `valid` to Block (like the dirty bit). Invalid operation just set the bit to false. Then we could reuse the Block's memory. We could consider this when you start to refactor the old system.

I think it's not good, because the invalidation is on file-level now. I would rather destroy a file-level HashMap to do the invalidation, than iterate it and mark all the blocks as invalid. Besides, the space of a `Block` is shared via an `Arc`, therefore we can reuse the sapce only if the block is not referenced outside of the cache.

Reply:
Reuse memory is complex. Destorying a file-level hashmap is a good idea.